### ISSUE TITLE
[RFC] Give back control of password hashing algorithm to app developer

### ISSUE THREAD

**Issue by teohhanhui** [symfony/symfony#33054]
**Description**  
(From https://github.com/symfony/symfony/issues/30914#issuecomment-518803478, https://github.com/symfony/symfony/issues/30914#issuecomment-518843060, https://github.com/symfony/symfony/issues/30914#issuecomment-518847109)

I think what's still very disturbing to me is that the app developer will completely lose control over which password hashing algorithm to use.

What's worse, if you have a scenario where you upgrade your PHP version / sodium extension and a new algorithm becomes supported, then if you rollback the app, the password would then fail to verify. There is no way to prevent this from happening if the app developer is forced to use `auto`.

TL;DR: `auto` is a good default, but choice is good (and sometimes necessary).

**Example**  
```yaml
encoders:
    App\Entity\User:
        algorithm: argon2id
```

**Comment by mdeboer** [comment]
Completely agree, if you want something that "just works" and don't care about these things, use Laravel.

**Comment by teohhanhui** [comment]
For some background on how this would be a problem, see https://github.com/symfony/symfony/issues/28093

**Comment by teohhanhui** [comment]
For an alternative to `auto`, see:

https://docs.spring.io/spring-security/site/docs/5.1.5.RELEASE/reference/html5/#pe-dpe
https://docs.spring.io/spring-security/site/docs/5.1.5.RELEASE/api/org/springframework/security/crypto/password/DelegatingPasswordEncoder.html
https://docs.spring.io/spring-security/site/docs/5.1.5.RELEASE/api/org/springframework/security/crypto/factory/PasswordEncoderFactories.html#createDelegatingPasswordEncoder--

(from https://github.com/symfony/symfony/pull/31019#issuecomment-482171738)

**Comment by nicolas-grekas** [comment]
How is this an alternative to auto? Spring recommends exactly the same as the new option to me.

**Comment by teohhanhui** [comment]
@nicolas-grekas The difference with the `DelegatingPasswordEncoder` in Spring is that the app developer is in control of which algorithms are being used. The `NativePasswordEncoder` in Symfony does not provide that control.

**Comment by nicolas-grekas** [comment]
See `MigratingPasswordEncoder`, which is the one that you should compare `DelegatingPasswordEncoder` too.

**Comment by teohhanhui** [comment]
But which are the encoders can we use with `MigratingPasswordEncoder`, when they're all deprecated in favour of `NativePasswordEncoder`?

For example, I want to be able to use `MigratingPasswordEncoder` with `argon2id` as the "best" encoder, and `argon2i` and `bcrypt` as "extra" encoders.

**Comment by stof** [comment]
@nicolas-grekas `MigratingPasswordEncoder` is fine, but still does not give you control over the algorithm being used. It gives you control over the encoder being used, but our non-deprecated encoders are not about an algorithm, but about an implementation.

**Comment by romaricdrigon** [comment]
I also think that using something else that `auto` should not trigger a deprecation warning.
On an application dealing with legacy users, you want to force the algorithm used. 
I don't see why Symfony should care about your data, to the point of raising deprecation warnings (well, at least not if you are using a hashing algorithm considered safe).

**Comment by nicolas-grekas** [comment]
Implementation wise, this should be a new constructor argument to native/sodium encoders.
Anyone up for a PR?

**Comment by chalasr** [comment]
I'm on it.
